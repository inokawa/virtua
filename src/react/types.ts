import { CSSProperties, ReactNode } from "react";

export type ViewportComponentAttributes = Pick<
  React.HTMLAttributes<HTMLElement>,
  "className" | "style" | "id" | "role" | "tabIndex" | "onKeyDown" | "onWheel"
> &
  React.AriaAttributes;

export interface CustomContainerComponentProps {
  style: CSSProperties;
  children: ReactNode;
}

export type CustomContainerComponent = React.ForwardRefExoticComponent<
  React.PropsWithoutRef<CustomContainerComponentProps> &
    React.RefAttributes<any>
>;

/**
 * Props of customized item component for {@link Virtualizer} or {@link WindowVirtualizer}.
 */
export interface CustomItemComponentProps {
  style: CSSProperties;
  index: number;
  children: ReactNode;
}

export type CustomItemComponent = React.ForwardRefExoticComponent<
  React.PropsWithoutRef<CustomItemComponentProps> & React.RefAttributes<any>
>;
