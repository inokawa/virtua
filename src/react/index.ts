export { VList } from "./VList.js";
export type { VListProps, VListHandle } from "./VList.js";
export { Virtualizer } from "./Virtualizer.js";
export type { VirtualizerProps, VirtualizerHandle } from "./Virtualizer.js";
export { WindowVirtualizer } from "./WindowVirtualizer.js";
export type {
  WindowVirtualizerProps,
  WindowVirtualizerHandle,
} from "./WindowVirtualizer.js";
export { VGrid as experimental_VGrid } from "./VGrid.js";
export type {
  VGridItemResize,
  VGridProps,
  VGridHandle,
  CustomCellComponent,
  CustomCellComponentProps,
} from "./VGrid.js";
export type * from "./types.js";
