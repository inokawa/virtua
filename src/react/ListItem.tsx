import {
  memo,
  useRef,
  useMemo,
  CSSProperties,
  ReactElement,
  ReactNode,
} from "react";
import { useIsomorphicLayoutEffect } from "./useIsomorphicLayoutEffect";
import { ItemResizeObserver } from "../core/resizer";
import { refKey } from "./utils";
import { isRTLDocument } from "../core/environment";
import { CustomItemComponent } from "./types";
import { NULL } from "../core/utils";

interface ListItemProps {
  _children: ReactNode;
  _resizer: ItemResizeObserver;
  _index: number;
  _offset: number;
  _hide: boolean;
  _as: "div" | CustomItemComponent;
  _isHorizontal: boolean;
  _isSSR: boolean | undefined;
}

/**
 * @internal
 */
export const ListItem = memo(
  ({
    _children: children,
    _resizer: resizer,
    _index: index,
    _offset: offset,
    _hide: hide,
    _as: Element,
    _isHorizontal: isHorizontal,
    _isSSR: isSSR,
  }: ListItemProps): ReactElement => {
    const ref = useRef<HTMLDivElement>(NULL);

    // The index may be changed if elements are inserted to or removed from the start of props.children
    useIsomorphicLayoutEffect(() => resizer(ref[refKey]!, index), [index]);

    const style = useMemo((): CSSProperties => {
      const style: CSSProperties = {
        position: hide && isSSR ? undefined : "absolute",
        [isHorizontal ? "height" : "width"]: "100%",
        [isHorizontal ? "top" : "left"]: 0,
        [isHorizontal ? (isRTLDocument() ? "right" : "left") : "top"]: offset,
        visibility: !hide || isSSR ? "visible" : "hidden",
      };
      if (isHorizontal) {
        style.display = "flex";
      }
      return style;
    }, [offset, hide, isSSR, isHorizontal]);

    if (typeof Element === "string") {
      return (
        <Element ref={ref} style={style}>
          {children}
        </Element>
      );
    } else {
      return (
        <Element ref={ref} style={style} index={index}>
          {children}
        </Element>
      );
    }
  }
);
