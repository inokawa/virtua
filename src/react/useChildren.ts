import { ReactElement, ReactNode, useMemo } from "react";
import { ItemElement, flattenChildren } from "./utils.js";

/**
 * @internal
 */
export const useChildren = <T>(
  children: ReactNode | ((data: T, i: number) => ReactElement),
  data: ArrayLike<T> | undefined
) => {
  return useMemo((): [(i: number) => ItemElement, number] => {
    if (typeof children === "function") {
      return [(i) => children(data![i]!, i), data!.length];
    }
    // Memoize element array
    const _elements = flattenChildren(children);
    return [(i) => _elements[i]!, _elements.length];
  }, [children, data]);
};
