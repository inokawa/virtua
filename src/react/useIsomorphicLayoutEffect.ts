import { useEffect, useLayoutEffect } from "react";
import { isBrowser } from "../core";

/**
 * https://gist.github.com/gaearon/e7d97cdf38a2907924ea12e4ebdf3c85
 * @internal
 */
export const useIsomorphicLayoutEffect = isBrowser
  ? useLayoutEffect
  : useEffect;
