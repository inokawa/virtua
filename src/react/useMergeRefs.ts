import { Ref, MutableRefObject, useCallback } from "react";

/**
 * @internal
 * Merge multiple refs into a single ref callback.
 */
export function useMergeRefs<T>(...refs: Array<Ref<T> | undefined>): Ref<T> {
  return useCallback((instance: T | null) => {
    for (const ref of refs) {
      ref && updateRef(ref, instance);
    }
  }, refs);
}

function updateRef<T>(ref: Ref<T>, instance: T | null): void {
  if (typeof ref === "function") {
    ref(instance);
  } else if (ref) {
    (ref as MutableRefObject<T | null>).current = instance;
  }
}
