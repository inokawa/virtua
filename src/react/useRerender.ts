import { useReducer } from "react";
import { VirtualStore } from "../core";

/**
 * @internal
 */
export const useRerender = (store: VirtualStore): (() => void) => {
  return useReducer(
    store._getStateVersion,
    undefined,
    store._getStateVersion
  )[1];
};
