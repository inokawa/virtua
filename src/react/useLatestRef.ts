import { useRef } from "react";
import { useIsomorphicLayoutEffect } from "./useIsomorphicLayoutEffect.js";
import { refKey } from "./utils.js";

/**
 * @internal
 */
export const useLatestRef = <T>(value: T) => {
  const ref = useRef<T>(value);

  useIsomorphicLayoutEffect(() => {
    ref[refKey] = value;
  }, [value]);

  return ref;
};
