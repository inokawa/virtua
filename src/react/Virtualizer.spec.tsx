import { afterEach, it, expect, vi } from "vitest";
import { render as _render, cleanup } from "@testing-library/react";
import { Virtualizer } from ".";
import { setupResizeJsDom } from "../../scripts/spec";

const ITEM_HEIGHT = 50;
const ITEM_WIDTH = 100;
const VIEWPORT_HEIGHT = ITEM_HEIGHT * 10;

setupResizeJsDom({
  itemSize: { width: ITEM_WIDTH, height: ITEM_HEIGHT },
  viewportSize: { width: ITEM_WIDTH, height: VIEWPORT_HEIGHT },
});

const render = (...args: Parameters<typeof _render>) => {
  const res = _render(...args);
  vi.runAllTicks();
  return res;
};

afterEach(cleanup);

it("should change components", () => {
  const { asFragment } = render(
    <div style={{ overflowY: "auto" }}>
      <Virtualizer as="ul" item="li">
        <div>0</div>
        <div>1</div>
        <div>2</div>
        <div>3</div>
        <div>4</div>
      </Virtualizer>
    </div>
  );
  expect(asFragment()).toMatchSnapshot();
});
