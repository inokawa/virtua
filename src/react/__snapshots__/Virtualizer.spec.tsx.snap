// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should change components 1`] = `
<DocumentFragment>
  <div
    style="overflow-y: auto;"
  >
    <ul
      style="overflow-anchor: none; flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 250px;"
    >
      <li
        style="position: absolute; width: 100%; left: 0px; top: 0px; visibility: visible;"
      >
        <div>
          0
        </div>
      </li>
      <li
        style="position: absolute; width: 100%; left: 0px; top: 50px; visibility: visible;"
      >
        <div>
          1
        </div>
      </li>
      <li
        style="position: absolute; width: 100%; left: 0px; top: 100px; visibility: visible;"
      >
        <div>
          2
        </div>
      </li>
      <li
        style="position: absolute; width: 100%; left: 0px; top: 150px; visibility: visible;"
      >
        <div>
          3
        </div>
      </li>
      <li
        style="position: absolute; width: 100%; left: 0px; top: 200px; visibility: visible;"
      >
        <div>
          4
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
